File: preprocessing/dag.py
import networkx as nx
import matplotlib.pyplot as plt
import pandas as pd

# Read edges from CSV
csv_file = '../quickest_path/data/USA-roads_sorted.csv'
edges_df = pd.read_csv(csv_file, names=['source', 'target', 'weight'])

def log_cycles(G):
    """
    Logs any cycles found in the directed graph.
    """
    try:
        cycles = list(nx.find_cycle(G, orientation='original'))
        if cycles:
            print("Cycles detected in the graph:")
            for cycle in cycles:
                cycle_nodes = " -> ".join(str(node) for node in cycle)
                print(f"Cycle found: {cycle_nodes}")
        else:
            print("No cycles detected. The graph is a DAG.")
    except nx.NetworkXNoCycle:
        print("No cycles detected. The graph is a DAG.")

def validate_dag(G):
    """
    Validates if the graph is a DAG.
    """
    if nx.is_directed_acyclic_graph(G):
        print("The graph is a valid DAG.")
    else:
        print("The graph contains cycles and is not a DAG.")



def check_connectivity(G):
    """
    Checks if the DAG is weakly connected without using library functions.
    If not, returns any separate components.
    """
    # Build an undirected adjacency list
    adjacency = {}
    for node in G.nodes():
        adjacency[node] = set()
    for source, target in G.edges():
        adjacency[source].add(target)
        adjacency[target].add(source)

    visited = set()
    components = []

    # BFS to identify connected components
    for node in G.nodes():
        if node not in visited:
            queue = [node]
            visited.add(node)
            component = {node}

            while queue:
                current = queue.pop(0)
                for neighbor in adjacency[current]:
                    if neighbor not in visited:
                        visited.add(neighbor)
                        queue.append(neighbor)
                        component.add(neighbor)

            components.append(component)

    if len(components) == 1:
        print("All nodes are connected in one main component.")
    else:
        print("The graph has multiple subgraphs:")
        for i, comp in enumerate(components, start=1):
            print(f"Subgraph {i}: {comp}")


# Ensure the CSV contains the correct columns
if 'source' in edges_df.columns and 'target' in edges_df.columns:
    # Create a directed graph
    G = nx.DiGraph()

    # Add edges from the CSV
    edges = list(zip(edges_df['source'], edges_df['target']))
    G.add_edges_from(edges)

    # Log and validate cycles
    # log_cycles(G)
    # validate_dag(G)
    check_connectivity(G)

else:
    print("CSV file must contain 'source' and 'target' columns.")

File: src/dag.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <sstream>

struct Edge {
    int source;
    int target;
};

std::vector<Edge> readCSV(const std::string& filePath) {
    std::vector<Edge> edges;
    std::ifstream file(filePath);
    std::string line;

    if (!file.is_open()) {
        std::cerr << "Error: Could not open file " << filePath << "\n";
        return edges;
    }

    while (std::getline(file, line)) {
        std::stringstream ss(line);
        std::string sourceStr, targetStr;
        if (!std::getline(ss, sourceStr, ',')) continue;
        if (!std::getline(ss, targetStr, ',')) continue;

        try {
            int source = std::stoi(sourceStr);
            int target = std::stoi(targetStr);
            edges.push_back({source, target});
        } catch (const std::exception& e) {
            std::cerr << "Warning: Failed to parse line: " << line << "\n";
        }
    }

    file.close();
    return edges;
}

void checkConnectivity(const std::vector<Edge>& edges) {
    std::unordered_map<int, std::unordered_set<int>> adjacency;
    for (const auto& edge : edges) {
        adjacency[edge.source].insert(edge.target);
        adjacency[edge.target].insert(edge.source);
    }

    std::unordered_set<int> visited;
    std::vector<std::unordered_set<int>> components;

    for (const auto& node : adjacency) {
        if (visited.find(node.first) == visited.end()) {
            std::queue<int> queue;
            std::unordered_set<int> component;
            queue.push(node.first);
            visited.insert(node.first);
            component.insert(node.first);

            while (!queue.empty()) {
                int current = queue.front();
                queue.pop();

                for (const auto& neighbor : adjacency[current]) {
                    if (visited.find(neighbor) == visited.end()) {
                        visited.insert(neighbor);
                        queue.push(neighbor);
                        component.insert(neighbor);
                    }
                }
            }

            components.push_back(component);
        }
    }

    if (components.size() == 1) {
        std::cout << "All nodes are connected in one main component.\n";
    } else {
        std::cout << "The graph has multiple subgraphs:\n";
        for (size_t i = 0; i < components.size(); ++i) {
            std::cout << "Subgraph " << i + 1 << ": ";
            for (const auto& node : components[i]) {
                std::cout << node << " ";
            }
            std::cout << "\n";
        }
    }
}

int main() {
    std::string filePath = "../data/USA-roads.csv";
    std::vector<Edge> edges = readCSV(filePath);

    if (!edges.empty()) {
        checkConnectivity(edges);
    } else {
        std::cerr << "CSV file must contain 'source' and 'target' columns.\n";
    }

    return 0;
}
